File: Source/Server/Game/Objects/Projectile.vb

Imports Mirage.Sharp.Asfw
Imports Core
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq

Friend Module Projectile
#Region "Type"

    Public Structure ProjectileRec
        Dim Name As String
        Dim Sprite As Integer
        Dim Range As Byte
        Dim Speed As Integer
        Dim Damage As Integer
    End Structure

    Public Structure MapProjectileRec
        Dim ProjectileNum As Integer
        Dim Owner As Integer
        Dim OwnerType As Byte
        Dim X As Integer
        Dim Y As Integer
        Dim Dir As Byte
        Dim Timer As Integer
    End Structure

#End Region

#Region "Database"
    Sub SaveProjectile(projectileNum As Integer)
        Dim json As String = JsonConvert.SerializeObject(Type.Projectile(projectileNum)).ToString()

        If RowExists(projectileNum, "projectile")
            UpdateRow(projectileNum, json, "projectile", "data")
        Else
            InsertRow(projectileNum, json, "projectile")
        End If
    End Sub

    Sub LoadProjectiles()
        Dim i As Integer

        For i = 1 To MAX_PROJECTILES
            LoadProjectile(i)
        Next
    End Sub

    Sub LoadProjectile(projectileNum As Integer)
        Dim data As JObject

        data = SelectRow(projectileNum, "projectile", "data")

        If data Is Nothing Then
            ClearProjectile(projectileNum)
            Exit Sub
        End If

        Dim projectileData = JObject.FromObject(data).toObject(Of ProjectileStruct)()
        Type.Projectile(projectileNum) = projectileData
    End Sub

    Sub ClearMapProjectile()
        Dim x As Integer
        Dim y As Integer

        ReDim MapProjectile(MAX_MAPS, MAX_PROJECTILES)

        For x = 1 To MAX_MAPS
            For y = 1 To MAX_PROJECTILES
                ClearMapProjectile(x, y)
            Next
        Next

    End Sub

    Sub ClearMapProjectile(MapNum As Integer, index As Integer)

        MapProjectile(MapNum, index).ProjectileNum = 0
        MapProjectile(MapNum, index).Owner = 0
        MapProjectile(MapNum, index).OwnerType = 0
        MapProjectile(MapNum, index).X = 0
        MapProjectile(MapNum, index).Y = 0
        MapProjectile(MapNum, index).Dir = 0
        MapProjectile(MapNum, index).Timer = 0

    End Sub

    Sub ClearProjectile(index As Integer)

        Type.Projectile(index).Name = ""
        Type.Projectile(index).Sprite = 0
        Type.Projectile(index).Range = 0
        Type.Projectile(index).Speed = 0
        Type.Projectile(index).Damage = 0

    End Sub

    Sub ClearProjectile()
        Dim i As Integer

        ReDim Type.Projectile(MAX_PROJECTILES)

        For i = 1 To MAX_PROJECTILES
            ClearProjectile(i)
        Next

    End Sub

#End Region

#Region "Incoming"

    Sub HandleRequestEditProjectile(index As Integer, ByRef data() As Byte)
        Dim buffer As New ByteStream(4)

        ' Prevent hacking
        If GetPlayerAccess(index) < AccessType.Developer Then Exit Sub
        If TempPlayer(index).Editor > -1 Then  Exit Sub

        Dim user As String

        user = IsEditorLocked(index, EditorType.Projectile)

        If user <> "" Then 
            PlayerMsg(index, "The game editor is locked and being used by " + user + ".", ColorType.BrightRed)
            Exit Sub
        End If

        TempPlayer(index).Editor = EditorType.Projectile

        buffer.WriteInt32(ServerPackets.SProjectileEditor)

        Socket.SendDataTo(index, buffer.Data, buffer.Head)
        buffer.Dispose()

    End Sub

    Sub HandleSaveProjectile(index As Integer, ByRef data() As Byte)
        Dim ProjectileNum As Integer
        Dim buffer As New ByteStream(data)

        If GetPlayerAccess(index) < AccessType.Developer Then Exit Sub

        ProjectileNum = buffer.ReadInt32

        ' Prevent hacking
        If Projectilenum <= 0 Or ProjectileNum > MAX_PROJECTILES Then
            Exit Sub
        End If

        Type.Projectile(ProjectileNum).Name = buffer.ReadString
        Type.Projectile(ProjectileNum).Sprite = buffer.ReadInt32
        Type.Projectile(ProjectileNum).Range = buffer.ReadInt32
        Type.Projectile(ProjectileNum).Speed = buffer.ReadInt32
        Type.Projectile(ProjectileNum).Damage = buffer.ReadInt32

        ' Save it
        SendUpdateProjectileToAll(ProjectileNum)
        SaveProjectile(ProjectileNum)
        Addlog(GetPlayerLogin(index) & " saved Projectile #" & ProjectileNum & ".", ADMIN_LOG)
        buffer.Dispose()

    End Sub

    Sub HandleRequestProjectile(index As Integer, ByRef data() As Byte)
        SendProjectiles(index)
    End Sub

    Sub HandleClearProjectile(index As Integer, ByRef data() As Byte)
        Dim ProjectileNum As Integer
        Dim Targetindex As Integer
        Dim TargetType As TargetType
        Dim TargetZone As Integer
        Dim mapNum As Integer
        Dim Damage As Integer
        Dim armor As Integer
        Dim npcnum As Integer
        Dim buffer As New ByteStream(data)
        ProjectileNum = buffer.ReadInt32
        Targetindex = buffer.ReadInt32
        TargetType = buffer.ReadInt32
        TargetZone = buffer.ReadInt32
        buffer.Dispose()

        mapNum = GetPlayerMap(index)

        Select Case MapProjectile(MapNum, ProjectileNum).OwnerType
            Case TargetType.Player
               If Type.MapProjectile(MapNum, ProjectileNum).Owner = index Then
                    Select Case TargetType
                        Case TargetType.Player

                            If IsPlaying(Targetindex) Then
                                If Targetindex <> index Then
                                    If CanPlayerAttackPlayer(index, Targetindex, True) = 1 Then

                                        ' Get the damage we can do
                                        Damage = GetPlayerDamage(index) + Type.Projectile(Type.MapProjectile(MapNum, ProjectileNum).ProjectileNum).Damage

                                        ' if the npc blocks, take away the block amount
                                        armor = CanPlayerBlockHit(Targetindex)
                                        Damage = Damage - armor

                                        ' randomise for up to 10% lower than max hit
                                        Damage = Random.NextDouble(1, Damage)

                                        AttackPlayer(index, Targetindex, Damage)
                                    End If
                                End If
                            End If

                        Case TargetType.NPC
                            npcnum = MapNPC(MapNum).NPC(Targetindex).Num
                            If CanPlayerAttackNpc(index, Targetindex, True) = 1 Then
                                ' Get the damage we can do
                                Damage = GetPlayerDamage(index) + Type.Projectile(Type.MapProjectile(MapNum, ProjectileNum).ProjectileNum).Damage

                                ' if the npc blocks, take away the block amount
                                armor = 0
                                Damage = Damage - armor

                                ' randomise from 1 to max hit
                                Damage = Random.NextDouble(1, Damage)

                                PlayerAttackNpc(index, Targetindex, Damage)
                            End If
                    End Select
                End If

        End Select

        ClearMapProjectile(MapNum, ProjectileNum)

    End Sub

#End Region

#Region "Outgoing"

    Sub SendUpdateProjectileToAll(ProjectileNum As Integer)
        Dim buffer As ByteStream

        buffer = New ByteStream(4)

        buffer.WriteInt32(ServerPackets.SUpdateProjectile)
        buffer.WriteInt32(ProjectileNum)
        buffer.WriteString((Type.Projectile(ProjectileNum).Name))
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Sprite)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Range)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Speed)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Damage)

        SendDataToAll(buffer.Data, buffer.Head)
        buffer.Dispose()

    End Sub

    Sub SendUpdateProjectileTo(index As Integer, ProjectileNum As Integer)
        Dim buffer As ByteStream

        buffer = New ByteStream(4)

        buffer.WriteInt32(ServerPackets.SUpdateProjectile)
        buffer.WriteInt32(ProjectileNum)
        buffer.WriteString((Type.Projectile(ProjectileNum).Name))
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Sprite)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Range)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Speed)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Damage)

        Socket.SendDataTo(index, buffer.Data, buffer.Head)
        buffer.Dispose()

    End Sub

    Sub SendProjectiles(index As Integer)
        Dim i As Integer

        For i = 1 To MAX_PROJECTILES
            If Len(Type.Projectile(i).Name) > 0 Then
                Call SendUpdateProjectileTo(index, i)
            End If
        Next

    End Sub

    Sub SendProjectileToMap(MapNum As Integer, ProjectileNum As Integer)
        Dim buffer As ByteStream

        buffer = New ByteStream(4)
        buffer.WriteInt32(ServerPackets.SMapProjectile)

        With MapProjectile(MapNum, ProjectileNum)
            buffer.WriteInt32(ProjectileNum)
            buffer.WriteInt32(.ProjectileNum)
            buffer.WriteInt32(.Owner)
            buffer.WriteInt32(.OwnerType)
            buffer.WriteInt32(.Dir)
            buffer.WriteInt32(.X)
            buffer.WriteInt32(.Y)
        End With

        SendDataToMap(MapNum, buffer.Data, buffer.Head)
        buffer.Dispose()

    End Sub

#End Region

#Region "Functions"

    Friend Sub PlayerFireProjectile(index As Integer, Optional IsSkill As Integer = 0)
        Dim ProjectileSlot As Integer
        Dim ProjectileNum As Integer
        Dim mapNum As Integer
        Dim i As Integer

        mapNum = GetPlayerMap(index)

        'Find a free projectile
        For i = 1 To MAX_PROJECTILES
           If Type.MapProjectile(MapNum, i).ProjectileNum = 0 Then ' Free Projectile
                ProjectileSlot = i
                Exit For
            End If
        Next

        'Check for no projectile, if so just overwrite the first slot
        If ProjectileSlot = 0 Then ProjectileSlot = 1

        'Check for skill, if so then load data acordingly
        If IsSkill > 0 Then
            ProjectileNum = Type.Skill(IsSkill).Projectile
        Else
            ProjectileNum = Type.Item(GetPlayerEquipment(index, EquipmentType.Weapon)).Projectile
        End If

        If ProjectileNum = 0 Then Exit Sub

        With MapProjectile(MapNum, ProjectileSlot)
            .ProjectileNum = ProjectileNum
            .Owner = index
            .OwnerType = TargetType.Player
            .Dir = GetPlayerDir(index)
            .X = GetPlayerX(index)
            .Y = GetPlayerY(index)
            .Timer = GetTimeMs() + 60000
        End With

        SendProjectileToMap(MapNum, ProjectileSlot)

    End Sub

    Friend Function Engine_GetAngle(CenterX As Integer, CenterY As Integer, targetX As Integer, targetY As Integer) As Single
        '************************************************************
        'Gets the angle between two points in a 2d plane
        '************************************************************
        Dim SideA As Single
        Dim SideC As Single

        On Error GoTo ErrOut

        'Check for horizontal lines (90 or 270 degrees)
        If CenterY = targetY Then
            'Check for going right (90 degrees)
            If CenterX < targetX Then
                Engine_GetAngle = 90
                'Check for going left (270 degrees)
            Else
                Engine_GetAngle = 270
            End If

            'Exit the function
            Exit Function
        End If

        'Check for horizontal lines (360 or 180 degrees)
        If CenterX = targetX Then
            'Check for going up (360 degrees)
            If CenterY > targetY Then
                Engine_GetAngle = 360

                'Check for going down (180 degrees)
            Else
                Engine_GetAngle = 180
            End If

            'Exit the function
            Exit Function
        End If

        'Calculate Side C
        SideC = Math.Sqrt(Math.Abs(targetX - CenterX) ^ 2 + Math.Abs(targetY - CenterY) ^ 2)

        'Side B = CenterY

        'Calculate Side A
        SideA = Math.Sqrt(Math.Abs(targetX - CenterX) ^ 2 + targetY ^ 2)

        'Calculate the angle
        Engine_GetAngle = (SideA ^ 2 - CenterY ^ 2 - SideC ^ 2) / (CenterY * SideC * -2)
        Engine_GetAngle = (Math.Atan(-Engine_GetAngle / Math.Sqrt(-Engine_GetAngle * Engine_GetAngle + 1)) + 1.5708) * 57.29583

        'If the angle is >180, subtract from 360
        If targetX < CenterX Then Engine_GetAngle = 360 - Engine_GetAngle

        'Exit function
        Exit Function

        'Check for error
ErrOut:

        'Return a 0 saying there was an error
        Engine_GetAngle = 0

        Exit Function
    End Function

#End Region

End Module

File: Source/Server/Game/Objects/Projectile.cs
using System;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using Mirage.Sharp.Asfw;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using static Core.Type;
using static Core.Global.Command;
using static Core.Packets;
using Core;

namespace Server
{

    public class Projectile
    {

        #region Database
        public static void SaveProjectile(int projectileNum)
        {
            string json = JsonConvert.SerializeObject(Data.Projectile[projectileNum]).ToString();

            if (Database.RowExists(projectileNum, "projectile"))
            {
                Database.UpdateRow(projectileNum, json, "projectile", "data");
            }
            else
            {
                Database.InsertRow(projectileNum, json, "projectile");
            }
        }

        public static async System.Threading.Tasks.Task LoadProjectilesAsync()
        {
            int i;

            var loopTo = Core.Constant.MAX_PROJECTILES;
            for (i = 0; i < loopTo; i++)
                await LoadProjectile(i);
        }

        public static async System.Threading.Tasks.Task LoadProjectile(int projectileNum)
        {
            JObject data;

            data = await Database.SelectRowAsync(projectileNum, "projectile", "data");

            if (data is null)
            {
                ClearProjectile(projectileNum);
                return;
            }

            var projectileData = data.ToObject<Core.Type.Projectile>();
            Core.Data.Projectile[projectileNum] = projectileData;
        }

        public static void ClearMapProjectile(int mapNum, int index)
        {

            Data.MapProjectile[mapNum, index].ProjectileNum = 0;
            Data.MapProjectile[mapNum, index].Owner = 0;
            Data.MapProjectile[mapNum, index].OwnerType = 0;
            Data.MapProjectile[mapNum, index].X = 0;
            Data.MapProjectile[mapNum, index].Y = 0;
            Data.MapProjectile[mapNum, index].Dir = 0;
            Data.MapProjectile[mapNum, index].Timer = 0;

        }

        public static void ClearProjectile(int index)
        {

            Data.Projectile[index].Name = "";
            Data.Projectile[index].Sprite = 0;
            Data.Projectile[index].Range = 0;
            Data.Projectile[index].Speed = 0;
            Data.Projectile[index].Damage = 0;

        }

        #endregion

        #region Incoming

        public static void HandleRequestEditProjectile(int index, ref byte[] data)
        {
            var buffer = new ByteStream(4);

            // Prevent hacking
            if (GetPlayerAccess(index) < (byte)AccessLevel.Developer)
                return;

            string user;

            user = IsEditorLocked(index, (byte)EditorType.Projectile);

            if (!string.IsNullOrEmpty(user))
            {
                NetworkSend.PlayerMsg(index, "The game editor is locked and being used by " + user + ".", (int) Color.BrightRed);
                return;
            }

            SendProjectiles(index);

            Core.Data.TempPlayer[index].Editor = (byte)EditorType.Projectile;

            buffer.WriteInt32((int) ServerPackets.SProjectileEditor);

            NetworkConfig.Socket.SendDataTo(index, buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();

        }

        public static void HandleSaveProjectile(int index, ref byte[] data)
        {
            int ProjectileNum;
            var buffer = new ByteStream(data);

            if (GetPlayerAccess(index) < (byte)AccessLevel.Developer)
                return;

            ProjectileNum = buffer.ReadInt32();

            // Prevent hacking
            if (ProjectileNum < 0 | ProjectileNum > Core.Constant.MAX_PROJECTILES)
            {
                return;
            }

            Data.Projectile[ProjectileNum].Name = buffer.ReadString();
            Data.Projectile[ProjectileNum].Sprite = buffer.ReadInt32();
            Data.Projectile[ProjectileNum].Range = (byte)buffer.ReadInt32();
            Data.Projectile[ProjectileNum].Speed = buffer.ReadInt32();
            Data.Projectile[ProjectileNum].Damage = buffer.ReadInt32();

            // Save it
            SendUpdateProjectileToAll(ProjectileNum);
            SaveProjectile(ProjectileNum);
            Core.Log.Add(GetAccountLogin(index) + " saved Projectile #" + ProjectileNum + ".", Constant.ADMIN_LOG);
            buffer.Dispose();

        }

        public static void HandleRequestProjectile(int index, ref byte[] data)
        {
            int ProjectileNum;

            var buffer = new ByteStream(data);
            ProjectileNum = buffer.ReadInt32();
            buffer.Dispose();

            SendProjectile(index, ProjectileNum);
        }

        public static void HandleClearProjectile(int index, ref byte[] data)
        {
            int ProjectileNum;
            int Targetindex;
            Core.TargetType TargetType;
            int TargetZone;
            int mapNum;
            int Damage;
            int armor;
            int NpcNum;
            var buffer = new ByteStream(data);
            ProjectileNum = buffer.ReadInt32();
            Targetindex = buffer.ReadInt32();
            TargetType = (Core.TargetType)buffer.ReadInt32();
            TargetZone = buffer.ReadInt32();
            buffer.Dispose();

            mapNum = GetPlayerMap(index);

            ClearMapProjectile(mapNum, ProjectileNum);

        }

        #endregion

        #region Outgoing

        public static void SendUpdateProjectileToAll(int ProjectileNum)
        {
            ByteStream buffer;

            buffer = new ByteStream(4);

            buffer.WriteInt32((int) ServerPackets.SUpdateProjectile);
            buffer.WriteInt32(ProjectileNum);
            buffer.WriteString(Data.Projectile[ProjectileNum].Name);
            buffer.WriteInt32(Data.Projectile[ProjectileNum].Sprite);
            buffer.WriteInt32(Data.Projectile[ProjectileNum].Range);
            buffer.WriteInt32(Data.Projectile[ProjectileNum].Speed);
            buffer.WriteInt32(Data.Projectile[ProjectileNum].Damage);

            NetworkConfig.SendDataToAll(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();

        }

        public static void SendUpdateProjectileTo(int index, int ProjectileNum)
        {
            ByteStream buffer;

            buffer = new ByteStream(4);

            buffer.WriteInt32((int) ServerPackets.SUpdateProjectile);
            buffer.WriteInt32(ProjectileNum);
            buffer.WriteString(Data.Projectile[ProjectileNum].Name);
            buffer.WriteInt32(Data.Projectile[ProjectileNum].Sprite);
            buffer.WriteInt32(Data.Projectile[ProjectileNum].Range);
            buffer.WriteInt32(Data.Projectile[ProjectileNum].Speed);
            buffer.WriteInt32(Data.Projectile[ProjectileNum].Damage);

            NetworkConfig.Socket.SendDataTo(index, buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();

        }

        public static void SendProjectile(int index, int projectileNum)
        {
            SendUpdateProjectileTo(index, projectileNum);
        }

        public static void SendProjectiles(int index)
        {
            var loopTo = Core.Constant.MAX_PROJECTILES;
            for (int i = 0; i < loopTo; i++)
            {
                if (Strings.Len(Data.Projectile[i].Name) > 0)
                {
                    SendUpdateProjectileTo(index, i);
                }
            }

        }

        public static void SendProjectileToMap(int mapNum, int ProjectileNum)
        {
            ByteStream buffer;

            buffer = new ByteStream(4);
            buffer.WriteInt32((int) ServerPackets.SMapProjectile);

            var withBlock = Data.MapProjectile[mapNum, ProjectileNum];
            buffer.WriteInt32(ProjectileNum);
            buffer.WriteInt32(withBlock.ProjectileNum);
            buffer.WriteInt32(withBlock.Owner);
            buffer.WriteInt32(withBlock.OwnerType);
            buffer.WriteInt32(withBlock.Dir);
            buffer.WriteInt32(withBlock.X);
            buffer.WriteInt32(withBlock.Y);          

            NetworkConfig.SendDataToMap(mapNum, buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();

        }

        #endregion

        #region Functions

        public static void PlayerFireProjectile(int index, int IsSkill = 0)
        {
            var ProjectileSlot = default(int);
            int ProjectileNum;
            int mapNum;
            int i;

            mapNum = GetPlayerMap(index);

            // Find a free projectile
            var loopTo = Core.Constant.MAX_PROJECTILES;
            for (i = 0; i < loopTo; i++)
            {
                if (Data.MapProjectile[mapNum, i].ProjectileNum == -1) // Free Projectile
                {
                    ProjectileSlot = i;
                    break;
                }
            }

            // Check for skill, if so then load data acordingly
            if (IsSkill > 0)
            {
                ProjectileNum = Data.Skill[IsSkill].Projectile;
            }
            else
            {
                ProjectileNum = Core.Data.Item[(int)GetPlayerEquipment(index, Equipment.Weapon)].Projectile;
            }

            if (ProjectileNum == -1)
                return;

            {
                var withBlock = Data.MapProjectile[mapNum, ProjectileSlot];
                withBlock.ProjectileNum = ProjectileNum;
                withBlock.Owner = index;
                withBlock.OwnerType = (byte)TargetType.Player;
                withBlock.Dir = (byte)GetPlayerDir(index);
                withBlock.X = GetPlayerX(index);
                withBlock.Y = GetPlayerY(index);
                withBlock.Timer = General.GetTimeMs() + 60000;
            }

            SendProjectileToMap(mapNum, ProjectileSlot);

        }

        public static float Engine_GetAngle(int CenterX, int CenterY, int targetX, int targetY)
        {
            float Engine_GetAngleRet = default;
            // ************************************************************
            // Gets the angle between two points in a 2d plane
            // ************************************************************
            float SideA;
            float SideC;
            try
            {

                // Check for horizontal lines (90 or 270 degrees)
                if (CenterY == targetY)
                {
                    // Check for going right (90 degrees)
                    if (CenterX < targetX)
                    {
                        Engine_GetAngleRet = 90f;
                    }
                    // Check for going left (270 degrees)
                    else
                    {
                        Engine_GetAngleRet = 270f;
                    }

                    // Exit the function
                    return Engine_GetAngleRet;
                }

                // Check for horizontal lines (360 or 180 degrees)
                if (CenterX == targetX)
                {
                    // Check for going up (360 degrees)
                    if (CenterY > targetY)
                    {
                        Engine_GetAngleRet = 360f;
                    }

                    // Check for going down (180 degrees)
                    else
                    {
                        Engine_GetAngleRet = 180f;
                    }

                    // Exit the function
                    return Engine_GetAngleRet;
                }

                // Calculate Side C
                SideC = (float)Math.Sqrt(Math.Pow(Math.Abs(targetX - CenterX), 2d) + Math.Pow(Math.Abs(targetY - CenterY), 2d));

                // Side B = CenterY

                // Calculate Side A
                SideA = (float)Math.Sqrt(Math.Pow(Math.Abs(targetX - CenterX), 2d) + Math.Pow(targetY, 2d));

                // Calculate the angle
                Engine_GetAngleRet = (float)((Math.Pow((double)SideA, 2d) - Math.Pow(CenterY, 2d) - Math.Pow((double)SideC, 2d)) / (double)(CenterY * SideC * -2));
                Engine_GetAngleRet = (float)((Math.Atan((double)-Engine_GetAngleRet / Math.Sqrt((double)(-Engine_GetAngleRet * Engine_GetAngleRet + 1f))) + 1.5708d) * 57.29583d);

                // If the angle is >180, subtract from 360
                if (targetX < CenterX)
                    Engine_GetAngleRet = 360f - Engine_GetAngleRet;

                // Exit function

                // Check for error
                return Engine_GetAngleRet;
            }
            catch
            {


                // Return a 0 saying there was an error
                Engine_GetAngleRet = 0f;

                return Engine_GetAngleRet;
            }
        }

        #endregion

    }
}